html/js/product-search.js

```diff
@@ -33,56 +47,82 @@ function match_product_to_preferences (product, product_preferences) {
 			// Iterate over attributes
 			
 			$.each(attribute_group.attributes, function(key, attribute) {
+
+				var attribute_preference = product_preferences[attribute.id];
 				
-				if ((! product_preferences[attribute.id]) || (product_preferences[attribute.id] == "not_important")) {
+				if ((! attribute_preference) || (attribute_preference == "not_important")) {
 					// Ignore attribute
 					debug += attribute.id + " not_important" + "\n";
 				}
 				else {
+
+					var attribute_factor = preferences_factors[attribute_preference];
+					sum_of_factors += attribute_factor;
 					
 					if (attribute.status == "unknown") {
-						
-						// If the attribute is important or more, then mark the product unknown
-						// if the attribute is unknown (unless the product is already not matching)				
-						
-						if (status == "yes") {
-							status = "unknown";
+
+						sum_of_factors_for_unknown_attributes += attribute_factor;
+
+						// If the attribute is mandatory and the attribute status is unknown
+						// then mark the product status unknown
+
+						if ((attribute_preference == "mandatory") && (status == "compatible")) {
+							status = "unknown_match";
 						}
 					}
 					else {
 						
-						debug += attribute.id + " " + product_preferences[attribute.id] + " - match: " + attribute.match + "\n";
-					
-						if (product_preferences[attribute.id] == "important") {
-					
-							score += attribute.match;
-						}
-						else if (product_preferences[attribute.id] == "very_important") {
-							
-							score += attribute.match * 2;
-						}
-						else if (product_preferences[attribute.id] == "mandatory") {
+						debug += attribute.id + " " + attribute_preference + " - match: " + attribute.match + "\n";
 
-							score += attribute.match * 4;
+						score += attribute.match * attribute_factor;
 					
-							if (attribute.match <= 20) {
-								status = "no";
-							}
+						// Mandatory attribute with a very bad score (e.g. contains an allergen) -> status: does not match
+						if ((attribute_preference == "mandatory") && (attribute.match <= 10)) {
+							status = "does_not_match";
+						}
+						// Mandatory attribute with a bad score (e.g. may contain traces of an allergen) -> status: may not match
+						else if ((attribute_preference == "mandatory") && (attribute.match <= 50)
+							&& ((status = "match") || (status = "may-not-match"))) {
+							status = "may_not_match";
 						}
 					}
 					
-					product.match_attributes[product_preferences[attribute.id]].push(attribute);
+					product.match_attributes[attribute_preference].push(attribute);
 				}
 			});
 		});		
 	}
 	else {
 		// the product does not have the attribute_group field 
-		status = "unknown";
+		status = "unknown_match";
+	}
+
+	// Normalize the score from 0 to 100
+	score /= sum_of_factors;
```

Comment:
```suggestion
	if (status !== "does_not_match") {
	  // Normalize the score from 0 to 100
	  score /= sum_of_factors;
	else {
	  score = 0;
	}
```

Reply:
Also, I'm not sure why we don't have `"may_not_match"` and `"unknown_match"` in the condition.